channeldb/control_tower.go

```diff
@@ -0,0 +1,515 @@
+package channeldb
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/coreos/bbolt"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/routing/route"
+)
+
+var (
+	// ErrAlreadyPaid signals we have already paid this payment hash.
+	ErrAlreadyPaid = errors.New("invoice is already paid")
+
+	// ErrPaymentInFlight signals that payment for this payment hash is
+	// already "in flight" on the network.
+	ErrPaymentInFlight = errors.New("payment is in transition")
+
+	// ErrPaymentNotInitiated is returned  if payment wasn't initiated in
+	// switch.
+	ErrPaymentNotInitiated = errors.New("payment isn't initiated")
+
+	// ErrPaymentAlreadyCompleted is returned in the event we attempt to
+	// recomplete a completed payment.
+	ErrPaymentAlreadyCompleted = errors.New("payment is already completed")
+
+	// ErrPaymentAlreadyFailed is returned in the event we attempt to
+	// re-fail a failed payment.
+	ErrPaymentAlreadyFailed = errors.New("payment has already failed")
+
+	// ErrUnknownPaymentStatus is returned when we do not recognize the
+	// existing state of a payment.
+	ErrUnknownPaymentStatus = errors.New("unknown payment status")
+)
+
+// ControlTower tracks all outgoing payments made by the switch, whose primary
+// purpose is to prevent duplicate payments to the same payment hash. In
+// production, a persistent implementation is preferred so that tracking can
+// survive across restarts. Payments are transition through various payment
+// states, and the ControlTower interface provides access to driving the state
+// transitions.
+type ControlTower interface {
+	// ClearForTakeoff atomically checks that no inflight or completed
+	// payments exist for this payment hash. If none are found, this method
+	// atomically transitions the status for this payment hash as InFlight.
+	ClearForTakeoff(info *CreationInfo) error
+
+	Attempt(lntypes.Hash, *AttemptInfo) error
```

Comment:
Can `ClearForTakeoff` be removed and `Attempt` be converted into `Attempt(lntypes.Hash, *AttemptInfo, failIfExists bool) error`?

If `failIfExists` is true, the attempt call fails if the payment hash is already in the db, either as in flight or succeeded. SendPayment optimistically makes all preparations for sending the payment off to the switch. But if right before that it tries to persist the attempt by calling `Attempt` with `failIfExists` true and it fails, it can also return the error to the caller there. There seems to be no side effects except for increase the pid.

Reply:
Done ;)